File: Sample-ObjC/ServicesInfo.m
//
//  ServiceInfo.m
//  Sample-ObjC
//
//  Created by Stas Kochkin on 08.05.2020.
//  Copyright © 2020 com.appodeal. All rights reserved.
//

#import "ServicesInfo.h"


@interface ServicesInfo ()

@property (nonatomic, copy, readwrite) NSString *appodealApiKey;

@end

@implementation ServicesInfo

+ (instancetype)sharedInfo {
    static ServicesInfo *_sharedInfo;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInfo = [[ServicesInfo alloc] init];
    });
    return _sharedInfo;
}

- (instancetype)init {
    if (self = [super init]) {
        [self populateFromServiceInfoPlist];
    }
    return self;
}
 
- (void)populateFromServiceInfoPlist {
    NSString *path = [NSBundle.mainBundle pathForResource:@"Services-Info" ofType:@"plist"];
    NSData *xml = [NSFileManager.defaultManager contentsAtPath:path];
    NSDictionary *plist = [NSPropertyListSerialization propertyListWithData:xml
                                                                    options:NSPropertyListMutableContainersAndLeaves
                                                                     format:nil
                                                                      error:nil];
    self.appodealApiKey = plist[@"Appodeal"][@"ApiKey"];
}

@end

File: Sample-Swift/ServicesInfo.swift
//
//  ServicesInfo.swift
//  Sample-Swift
//
//  Created by Stas Kochkin on 08.05.2020.
//  Copyright © 2020 com.appodeal. All rights reserved.
//

import Foundation


let servicesInfo: ServicesInfo = {
    let xml: Data! = Bundle.main
        .path(forResource: "Services-Info", ofType: "plist")
        .flatMap(FileManager.default.contents)
    let decoder = PropertyListDecoder()
    return try! decoder.decode(
        ServicesInfo.self,
        from: xml
    )
}()


struct ServicesInfo: Decodable {
    enum Keys: String, CodingKey {
        case appodeal = "Appodeal"
    }
    
    struct Appodeal: Decodable {
        enum Keys: String, CodingKey {
            case apiKey = "ApiKey"
        }
        
        var apiKey: String
        
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: Keys.self)
            apiKey = try container.decode(String.self, forKey: .apiKey)
        }
    }
    
    var appodeal: Appodeal
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: Keys.self)
        appodeal = try container.decode(ServicesInfo.Appodeal.self, forKey: .appodeal)
    }
}
